### PR TITLE
Fix 21258 - Tuple parameters use the first element of a default tuple value

### PR DESCRIPTION
**Pull request by Geod24**
The logic tried to use a single variable for both keeping track of the
original index and the extended index, but actually only tracked the
extended index.
Since `if (tupleStartIdx < eidx) tupleStartIdx = eidx` would be true
for every iteration, we would always assign the value at index 0.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
@Geod24 very nice!

**Comment by Geod24** [comment]
Let's hold on for a couple hours, maybe someone will suggest additional tests (that someone might be me).

**Comment by UplinkCoder** [comment]
@Geod24 If you could improve the commit message, to say what the typo caused that'd be good.

**Comment by Geod24** [comment]
Heh, the logic was flawed anyway and failed for tuples > 2 elements. Should be properly fixed now.

**Comment by rmanthorpe** [comment]
Thanks for sorting it out so quickly!

File: src/dmd/typesem.d [hunk 1]
```diff
@@ -1494,7 +1494,10 @@ extern(C++) Type typeSemantic(Type type, const ref Loc loc, Scope* sc)
             // Now that we completed semantic for the argument types,
             // run semantic on their default values,
             // bearing in mind tuples have been expanded.
-            size_t tupleStartIdx = size_t.max;
+            // We need to keep a pair of [oidx, eidx] (original index,
+            // extended index), so we now to run semantic when the original
+            // index changes for the first time.
+            size_t[2] tupleStartIdx = [ size_t.max, size_t.max ];
```
Comment (WalterBright) [inline src/dmd/typesem.d:1500]: I don't see a purpose for making this an array, as it is never indexed. Might as well call it `tupleStartIdx1` and `tupleStartIdx2`. Or `tupleOrigIdx` and `tupleExtIdx`.

**Comment by Geod24** [comment]
Fixed: Changed static array to two variables.

File: src/dmd/typesem.d [hunk 2]
```diff
@@ -1494,7 +1494,11 @@ extern(C++) Type typeSemantic(Type type, const ref Loc loc, Scope* sc)
             // Now that we completed semantic for the argument types,
             // run semantic on their default values,
             // bearing in mind tuples have been expanded.
-            size_t tupleStartIdx = size_t.max;
+            // We need to keep a pair of [oidx, eidx] (original index,
+            // extended index), so we now to run semantic when the original
```
Comment (thewilsonator) [inline src/dmd/typesem.d:1498]: grammar. get rid of `to`.

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/typesem.d (hunk 2 above)
Comment (Geod24) [reply src/dmd/typesem.d]: Good catch. Actually the sentence wasn't quite accurate so I rephrased it.

**Comment by UplinkCoder** [comment]
Confirmed to fix our code.

**Pull request merged**